Rcar add support of mmc dma

xakep-amatop opened the pull request.

xakep-amatop committed:
drivers: rcar_mmc: create separate functions for dma and non-dma xref

split function rcar_mmc_rx_tx_data into two additional for
DMA and Non-DMA xref for better redability.

Signed-off-by: Mykola Kvach <mykola_kvach@epam.com>
Acked-by: Volodymyr Babchuk <volodymyr_babchuk@epam.com>
Acked-by: Dmytro Firsov <dmytro_firsov@epam.com>

xakep-amatop commented on drivers/sdhc/Kconfig.rcar:
```diff
@@ -12,7 +12,10 @@ if MMC_RCAR
-	default n
+	select CACHE_MANAGEMENT
+	select DCACHE
+	depends on !SHIELD_XENVM_DOM0
```
(note) will fix this when solve problem with cache set/way instructions

xakep-amatop replied:
ohhh, I understand it can be used not only for Dom0

xakep-amatop committed:
arch: arm64: mmu: WA: avoid using of set/way cache instructions for Xen

Unfortunately, set/way ops are not easily virtualized by Xen. S/W
emulation is disabled, because IP-MMU is active for Dom0. IP-MMU is
a IO-MMU made by Renesas, as any good IO-MMU, it shares page-tables
with CPU. Trying to emulate S/W with IP-MMU active will lead to
IO-MMU faults. So if we build Zephyr as a Xen Initial domain, it
won't work with cache management support enabled.

In this commit I use VA data invalidate instead of set/way --
invalidate all virtual address space, so it might increase system
initialization time.

Note: this patch is applied as a simple and fast workaround, we need
      to find more optimal solution.

Signed-off-by: Mykola Kvach <mykola_kvach@epam.com>
Acked-by: Volodymyr Babchuk <volodymyr_babchuk@epam.com>
Acked-by: Dmytro Firsov <dmytro_firsov@epam.com>

xakep-amatop committed:
drivers: rcar_mmc: enable DMA support for XenVM Initial domain

After implementing a workaraound:
 "arch: arm64: mmu: WA: avoid using of set/way cache instructions for Xen"
it is possible to enable Renesas MMC DMA support for Xen Initial domain,
e.g. Zephyr cache managment system can't work in a proper way with Xen
dom0.

Signed-off-by: Mykola Kvach <mykola_kvach@epam.com>
Acked-by: Volodymyr Babchuk <volodymyr_babchuk@epam.com>
Acked-by: Dmytro Firsov <dmytro_firsov@epam.com>

xakep-amatop committed:
drivers: rcar_mmc: use events from IRQ for DMA instead of polling

Add support of MMC IRQ and trigger event from IRQ handler instead of
polling DMA IRQ flags. It saves CPU cycles and other threads may use
CPU time for their purposes.

Signed-off-by: Mykola Kvach <mykola_kvach@epam.com>
Acked-by: Volodymyr Babchuk <volodymyr_babchuk@epam.com>
Acked-by: Dmytro Firsov <dmytro_firsov@epam.com>

lorc commented on drivers/sdhc/rcar_mmc.c:
```diff
@@ -451,15 +452,22 @@ static uint32_t rcar_mmc_gen_data_cmd(struct sdhc_command *cmd,
+	struct sdhc_data *data,
+	bool is_read)
+{
+	int ret = 0;
```
I'd expect to see `ret = -ENOSYS` , which means "Function not implemented". Which, in turn, exactly describe state of this function it the commit `drivers: rcar_mmc: create separate functions for dma and non-dma xref` .

xakep-amatop replied:
I've added error status only to the first commit in which the function hadn't implementation yet
we don't need this status anymore

lorc commented on drivers/sdhc/rcar_mmc.c:
```diff
@@ -469,26 +477,18 @@ static uint32_t rcar_mmc_gen_data_cmd(struct sdhc_command *cmd,
 	bool is_read)
 {
 	struct mmc_rcar_data *dev_data = dev->data;
-	uint32_t info1_reg;
```
Looks like you are doing more than commit description says...

xakep-amatop replied:
nope, it is directly related to splitting of the function, this code still exist inside `rcar_mmc_rx_tx_data` function

lorc commented on drivers/sdhc/rcar_mmc.c:
```diff
@@ -187,14 +197,16 @@ static int rcar_mmc_check_errors(const struct device *dev)
  * @param flag polling flag(s)
  * @param state state of flag(s) when we should stop polling
  * @param check_errors call @ref rcar_mmc_check_errors function or not
+ * @param check_dma_errors check if there is DMA errors inside info2
```
if there **are** DMA errors

lorc commented on drivers/sdhc/rcar_mmc.c:
```diff
@@ -452,11 +471,77 @@ static uint32_t rcar_mmc_gen_data_cmd(struct sdhc_command *cmd,
+
+	dma_addr = z_mem_phys_addr(data->data);
+	rcar_mmc_write_reg32(dev, RCAR_MMC_DMA_ADDR_L, dma_addr);
+	rcar_mmc_write_reg32(dev, RCAR_MMC_DMA_ADDR_H, 0);
```
It is bold of you to assume that all memory buffers will reside in lower 4GB part of physical address space. Gen3 chips have much more memory above 4GBs. So, put `dma_addr >> 32` there, instead of `0`. You will save much headache for someone in the future.

xakep-amatop replied:
Renesas MMC DMAC doesn't support 64-bit addresses
probably in case of 64-bit buffer address we should switch to non-dma mode

lorc replied:
Ah, I see. In this case you must check that `dma_addr` holds, in fact, a 32-bit address

xakep-amatop replied:
I've added a checking before call this function

lorc commented on drivers/sdhc/rcar_mmc.c:
```diff
@@ -452,11 +471,77 @@ static uint32_t rcar_mmc_gen_data_cmd(struct sdhc_command *cmd,
+						dma_info1_poll_flag, false, true);
+
+	if (is_read) {
+		if (sys_cache_data_invd_range(data->data, data->blocks * data->block_size) < 0) {
```
Actually,  you are not required to invalidate cache after read.  You already did this before the DMA operation. All device drivers assume that no one touches data buffers when drivers are working with them.

xakep-amatop replied:
I found info that I need an invalidation here, also the Linux Renesas MMC driver do the same
```
Invalidate: Before cpu tries to read a portion of memory updated by a device, the corresponding memory needs to be invalidated.
```

could you please share with me where did you find this info?
> All device drivers assume that no one touches data buffers when drivers are working with them.

lorc replied:
>  Linux Renesas MMC driver do the same

I bet it does not invalidate cache before DMA operation. Which makes sense.

On other hand, I don't see sense in invalidating cache twice.  You already invalidated cache lines for this buffer before the DMA operation start. How you can get cache entries for this buffer if CPU does not touch it?  

> could you please share with me where did you find this info?

This is sort of general knowledge. Reasoning is following: user code does not know in what exact moment driver (or hardware in case of DMA) will read/write from/to a buffer. There is no synchronization primitives for such cases. So, when user code provides buffer to a driver, it is expected that user code will not touch the buffer until driver return an operation result. 

If you want some more formal source, I can propose the following quote from Linux Device Drivers:

> Once a buffer has been mapped, it belongs to the device, not the processor. Until
the buffer has been unmapped, the driver should not touch its contents in any
way. Only after dma_unmap_single has been called is it safe for the driver to
access the contents of the buffer (with one exception that we see shortly).
Among other things, this rule implies that a buffer being written to a device cannot be mapped until it contains all the data to write. 

This is about DMA operations, but you got the idea.

Another point: `dma-buf` interface in Linux provides to clear functions: `begin_cpu_access` and `end_cpu_access` to control when CPU might access shared DMA buffers.

lorc commented on drivers/sdhc/rcar_mmc.c:
```diff
@@ -486,9 +571,6 @@ static int rcar_mmc_sd_buf_rx_tx_data(const struct device *dev,
 	int ret = 0;
 	uint32_t info2_poll_flag = is_read ? RCAR_MMC_INFO2_BRE : RCAR_MMC_INFO2_BWE;
 
-	rcar_mmc_write_reg32(dev, RCAR_MMC_INFO2_MASK,
```
Are those changes related to the DMA functionality?

xakep-amatop replied:
I just hid unnecessary MMC IRQ triggering, 
I'll move these changes to a separate commit

xakep-amatop replied:
moved to a separate commit

lorc commented on drivers/sdhc/rcar_mmc.c:
```diff
@@ -551,9 +633,6 @@ static int rcar_mmc_rx_tx_data(const struct device *dev,
 	uint32_t info1_reg;
 	int ret = 0;
 
-	rcar_mmc_write_reg32(dev, RCAR_MMC_INFO1_MASK,
```
Are those changes related to the DMA functionality?

lorc commented on drivers/sdhc/rcar_mmc.c:
```diff
@@ -616,9 +695,6 @@ static int rcar_mmc_request(const struct device *dev,
 
 	rcar_mmc_reset_and_mask_irqs(dev);
 
-	rcar_mmc_write_reg32(dev, RCAR_MMC_INFO1_MASK, (uint32_t)~RCAR_MMC_INFO1_RSP);
```
Are those changes related to the DMA functionality?

xakep-amatop replied:
moved to a separate commit

lorc commented on drivers/sdhc/rcar_mmc.c:
```diff
@@ -1596,13 +1674,17 @@ static int rcar_mmc_init(const struct device *dev)
+	 * note: we need to set thos flag before call rcar_mmc_init_controller_regs,
+	 *       because we want to cleanup dma regs and mask dma irqs
+	 */
+	data->dma_support = IS_ENABLED(CONFIG_RCAR_MMC_DMA_SUPPORT);
```
Why didn't you put this to the correct place in the previous patch?

xakep-amatop replied:
because only in this commit I add reset dma irqs and masks
I'll move this change to an appropriate commit

firscity replied:
Probably typo in note: thos -> those

xakep-amatop replied:
> Probably typo in note: thos -> those

fixed

xakep-amatop replied:
> Why didn't you put this to the correct place in the previous patch?

fixed

lorc commented on drivers/sdhc/rcar_mmc.c:
```diff
@@ -527,10 +538,33 @@ static int rcar_mmc_dma_rx_tx_data(const struct device *dev,
 	rcar_mmc_write_reg32(dev, RCAR_MMC_DMA_CTL, RCAR_MMC_DMA_CTL_START);
 
+	ret = k_sem_take(&dev_data->irq_xref_fin,
+			 K_USEC(MMC_POLL_FLAGS_TIMEOUT_US + 100 * data->blocks));
```
How you justify 100us per data block?

xakep-amatop replied:
it's just a some random value,
actually, I can measure a real value, but I suppose that I should measure it for different SD-IF clock frequencies

lorc commented on drivers/sdhc/rcar_mmc.c:
```diff
@@ -527,10 +538,33 @@ static int rcar_mmc_dma_rx_tx_data(const struct device *dev,
+	ret = k_sem_take(&dev_data->irq_xref_fin,
+			 K_USEC(MMC_POLL_FLAGS_TIMEOUT_US + 100 * data->blocks));
+	if (ret < 0) {
+		LOG_ERR("%s: interrupt signal timeout error %d", dev->name, ret);
```
This does not automatically mean that transaction failed. Maybe you not received IRQ for some other reason. You need to check your `INFO2` register anyways to see if where was any errors.

xakep-amatop replied:
added

lorc reviewed (commented):
You need to rewrite description for `arch: arm64: mmu: avoid using of set/way cache instructions for Xen`, because it is misleading.

1. Xen actually supports SET/WAY cache ops emulation. Please find `p2m_set_way_flush` function in Xen's `p2m.c` and read a huge comment above it for mode details. It is really interesting.
2. In this particular case SET/WAY emulation is disabled because IP-MMU is active for Dom0. IP-MMU is a IO-MMU made by Renesas, as any good IO-MMU, it shares page-tables with CPU. Long story short, trying to emulate SET/WAY with IP-MMU active will lead to IO-MMU faults. See https://patchwork.kernel.org/project/xen-devel/patch/alpine.DEB.2.21.2108031419500.19737@sstabellini-ThinkPad-T480s/ for more information.

So, it is incorrect to say that Xen does not support SET/WAY ops because they are old. Yes, they are old and are pain to emulate, but Xen tries its best to do this. Sadly, in some cases those instruction just can't be emulated without even more effort. 

Also "add some timeout during system
initialization"  does not sound correct.  How about "might increase system initialization time".   

Also "invalidate all VM" does not sound good too. You wanted to say "invalidate all virtual address space".

firscity commented on drivers/sdhc/rcar_mmc.c:
```diff
@@ -1613,6 +1647,33 @@ static int rcar_mmc_init_controller_regs(const struct device *dev)
+		rcar_mmc_write_reg32(dev, RCAR_MMC_DMA_INFO2_MASK, ~0);
+		k_sem_give(&data->irq_xref_fin);
+	} else {
+		LOG_WRN("%s: warning: non-dma event triggers irq", dev->name);
```
Be careful with printing in IRQ context, previously we've faced with some issues related to this. If it's not critical, I propose to omit this.

xakep-amatop replied:
Which kind of problems have you seen?

I have been debugging the IRQ handler using prints and haven't caught any issues during debug

Looks like `LOG_` API may work from IRQ context https://docs.zephyrproject.org/latest/services/logging/index.html:
```
Default frontend is engaged when logging API is called in a source of logging (e.g. LOG_INF) and is responsible for filtering a message (compile and run time), allocating buffer for the message, creating the message and committing that message. Since logging API can be called in an interrupt, frontend is optimized to log the message as fast as possible.
```

firscity commented on drivers/sdhc/rcar_mmc.c:
```diff
@@ -496,6 +504,9 @@ static int rcar_mmc_dma_rx_tx_data(const struct device *dev,
 	int ret = 0;
 	uint32_t dma_info1_poll_flag;
+#ifdef CONFIG_RCAR_MMC_DMA_IRQ_DRIVEN_SUPPORT
+	struct mmc_rcar_data *dev_data = dev->data;
```
Note: do we actually need this construction just to save 1 reference here?
`ret = k_sem_take(&dev_data->irq_xref_fin, ...`

xakep-amatop replied:
the data ptr inside dev is a pointer to a void, so it's not possible to dereference data ptr in the next way `dev->data->irq_xref_fin`

firscity commented on arch/arm64/core/mmu.c:
```diff
@@ -820,7 +820,12 @@ static void enable_mmu_el1(struct arm_mmu_ptables *ptables, unsigned int flags)
 	/* Invalidate all data caches before enable them */
+#ifdef CONFIG_DT_HAS_XEN_XEN_ENABLED
+	/* todo: find more optimal solution */
+	sys_cache_data_invd_range((void *)CONFIG_KERNEL_VM_BASE, CONFIG_KERNEL_VM_SIZE);
```
I think it will be great to add `WA: ` to the start of commit topic

xakep-amatop replied:
done

firscity commented on drivers/sdhc/Kconfig.rcar:
```diff
@@ -14,7 +14,6 @@ config RCAR_MMC_DMA_SUPPORT
 	bool "Internal DMA support for Renesas Rcar MMC driver"
 	select CACHE_MANAGEMENT
 	select DCACHE
-	depends on !SHIELD_XENVM_DOM0
```
Please, add more description for this commit ("We needed to turn of DMA due to cache..., now it has a workaround and we can...") after some time we will probably forget about a problem, that caused this changes.

firscity commented on drivers/sdhc/Kconfig.rcar:
```diff
@@ -23,6 +23,12 @@ if RCAR_MMC_DMA_SUPPORT
+	bool "Internal DMA IRQ driven support for Renesas Rcar MMC driver"
+	default y
+	help
+	  Internal DMA IRQ driven support for Renesas Rcar MMC driver.
```
I think that help line should have some explanation or rather be omitted here.

xakep-amatop committed:
drivers: rcar_mmc: add support of MMC DMA without Seq

Add all necessary configuration for proper work with MMC DMA.

An address of a data buffer has to be aligned to 128 bytes
if it is not, driver will use automatically NON-DMA mode.

Also Renesas MMC DMAC doesn't support 64-bit DMA addresses,
so for case when we have 64-bit xref data address we use
NON-DMA mode.

note: DMA isn't yet supported for case when we use XenVM Dom0
      shield, it is because Xen injects exception when it faces
      with Set/Way cache instructions and Zephyr has such
      instructions during init MMU.

Signed-off-by: Mykola Kvach <mykola_kvach@epam.com>
Acked-by: Volodymyr Babchuk <volodymyr_babchuk@epam.com>
Acked-by: Dmytro Firsov <dmytro_firsov@epam.com>

xakep-amatop committed:
drivers: rcar_mmc: disable all MMC IRQs

Disable all trigerrings of MMC IRQs, e.g. mask all IRQs,
because, currently, we don't use IRQ events from MMC.

Signed-off-by: Mykola Kvach <mykola_kvach@epam.com>
Acked-by: Volodymyr Babchuk <volodymyr_babchuk@epam.com>
Acked-by: Dmytro Firsov <dmytro_firsov@epam.com>

lorc commented on drivers/sdhc/rcar_mmc.c:
```diff
@@ -528,10 +540,34 @@ static int rcar_mmc_dma_rx_tx_data(const struct device *dev,
+	rcar_mmc_write_reg32(dev, RCAR_MMC_DMA_INFO1_MASK, reg);
 	rcar_mmc_write_reg32(dev, RCAR_MMC_DMA_CTL, RCAR_MMC_DMA_CTL_START);
 
+	ret = k_sem_take(&dev_data->irq_xref_fin,
```
I wonder what will happen if you'll get of sync in you semaphore usage. For example, there will be an extra DMA interrupt, of one of the interrupts will be late, so you'll return with timeout from ` k_sem_take`, but interrupt handler then will increase semaphore count. In the next operation, you'll return from `k_sem_take` immediately, which is wrong.

Linux in this case use "completions", which are free from such an issues. Or, in some cases, drivers check status register and restart wait for interrupt if transaction is not ready.

xakep-amatop replied:
good catch!

I have masked interrupts inside IRQ handler and IRQs have been masked before triggering of DMA start, so it isn't possible to have two or more IRQs in a row, but late arriving of the current interrupt is still possible, e.g. k_sem_take returns error code, I think one action will be enough here:
* reset the semaphore after resetting DMA IRQ masks.

lorc replied:
This is a bit unorthodox, but I can't see reason why this might now work.

lorc commented on drivers/sdhc/rcar_mmc.c:
```diff
@@ -462,11 +471,72 @@ static uint32_t rcar_mmc_gen_data_cmd(struct sdhc_command *cmd,
+		dma_info1_poll_flag = RCAR_MMC_DMA_INFO1_END_RD2;
+		reg |= RCAR_MMC_DMA_MODE_DIR_RD;
+	} else {
+		ret = sys_cache_data_flush_range(data->data, data->blocks * data->block_size);
```
We are discussed in DMs that flush should be called in either case. Becaue:

1. Before write (memory to device) operation we need CPU to finish all writes into data buffer, so DMA controller will see the same data as CPU.
2. Before read (device to memory) operation, we need to be sure that CPU will not spill cache back into the buffer, after DMA wrote something to it. Consider the following sequence of events: 
     1. user's code does `memset(buffer, 0)`,  this operation is cached, but not written to memory
     2.  user calls `read(buffer)`
     3. DMA writes data to `buffer`
     4. CPU decides that it is time to propage writes to the memory, so it fills `buffer` with zeros. 
     5. Whooops!

xakep-amatop replied:
done

lorc reviewed (commented):
> There are only two hard things in Computer Science: cache invalidation and naming things.
> 
> -- Phil Karlton

lorc reviewed (commented):
`Acked-by: Volodymyr Babchuk <volodymyr_babchuk@epam.com>`

firscity reviewed (approved):
`Acked-by: Dmytro Firsov <dmytro_firsov@epam.com>`
